lib/parse_identifier.rb

```diff
@@ -0,0 +1,121 @@
+require 'identifiers'
+
+# https://github.com/altmetric/identifiers
+#
+# Identifiers::DOI.extract('example: 10.123/abcd.efghi')
+# # => ["10.123/abcd.efghi"]
+#
+# It can support many identifiers:
+# Identifiers::AdsBibcode.extract('')
+# Identifiers::ArxivId.extract('')
+# Identifiers::DOI.extract('')
+# Identifiers::Handle.extract('')
+# Identifiers::ISBN.extract('')
+# Identifiers::NationalClinicalTrialId.extract('')
+# Identifiers::ORCID.extract('')
+# Identifiers::PubmedId.extract('')
+# Identifiers::RepecId.extract('')
+# Identifiers::URN.extract('')
+
+# Exception for incorrect instantiation of a parser (not necessarily bad data)
+class ParseIdentifierTypeError < StandardError
+
+end
+
+# Exception for bad data
+class ParseIdentifierInvalidError < StandardError
+
+end
+
+# Parse the identifiers in a PublicationIdentifier
+#
+# This is a base class for ParseIdentifier* subclasses.
+#
+# This class should try to work with the input parameter as a read-only object.
+# It should only modify it when explicitly asked to `update` it.  If it's
+# an active record object, assign new values to it, but don't save it, leave
+# that persistence action to the consumer; see the runner script in
+# @see script/publication_identifier_normalization.rb
+class ParseIdentifier
+
+  URI_PREFIX = nil # subclasses override this constant
+
+  attr_reader :pub_id
+  attr_reader :type
+
+  # @param pub_id [PublicationIdentifier]
+  def initialize(pub_id)
+    @pub_id = pub_id
+    @type = pub_id[:identifier_type]
+    raise(ParseIdentifierTypeError, "INVALID TYPE #{pub_id.inspect}") unless match_type
+    raise(ParseIdentifierInvalidError, "EMPTY DATA #{pub_id.inspect}") if empty?
+    raise(ParseIdentifierInvalidError, "INVALID DATA #{pub_id.inspect}") unless valid?
+  end
+
+  def empty?
+    pub_id[:identifier_value].blank? && pub_id[:identifier_uri].blank?
+  end
+
+  # Update the pub_id with parsed data
+  # @return pub_id [PublicationIdentifier]
+  def update
```

Comment:
I think `update` should not exist, in part because it gets around the checks you have tied to initialization, but also because it is redundant.  If the caller has a `PublicationIdentifier` and they want a new `ParseIdentifier` with that value, the call for that is `.new`.  

Consider, too, that all your instance variables that are cached up would become invalid.  `@value`, if previously accessed, would be the OLD `pub_id[:identifier_value]`.  It would NOT be repopulated.

Reply:
I don't think you understand how it works.  We should discuss it rather than try to type it out.  Somehow your concept of how it works is not the same as mine and I hope discussion can clarify the disconnect.

Reply:
Ah, you're right, I was misreading this as updating *this* object, but instead it is assigning inside `pub_id`.  Still not 100% clear to me, but not what I was thinking in my first comment.

Reply:
The `update` method can never be called if the initialization fails.